### PR TITLE
Messages home items

### LINKED ISSUE #14225: Messages home list new components
Currently messages home list items are using components that are modified "old components". Some parts of the new design are missing. New list item components need to be created in `quo2` for chat list items and contact list items to reflect the [new designs](https://www.figma.com/file/7KIYbhoqNGAIFonE0w9TDz/Messages-for-Mobile?node-id=1424%3A342207).

### PR DESCRIPTION
**Pull request by OmarBasem**
fixes #14225, #14267 

### Summary

This PR implements new messages home items and contact list items [new design](https://www.figma.com/file/7KIYbhoqNGAIFonE0w9TDz/Messages-for-Mobile?node-id=1424%3A342207).

status: ready

### CODE REVIEW

File: src/quo2/components/list_items/contact_item.cljs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            [quo2.foundations.typography :as typography]
+            [quo2.components.icon :as icons]
+            [quo2.foundations.colors :as colors]
+            [quo2.components.avatars.user-avatar :as user-avatar]
+            [quo.react-native :as rn]
+            [status-im.utils.utils :refer [get-shortened-address]]
+            [quo.platform :as platform]))
+
+(defn open-chat [chat-id]
+  (re-frame/dispatch [:dismiss-keyboard])
```
Comment (ilmotta) [inline src/quo2/components/list_items/contact_item.cljs]: I'm not sure anymore of the decision, but I remember @flexsurfer using `rf` as the require alias that we should be using. I like it because re-frame is so pervasive in the codebase that I think everybody can quickly internalize what `rf` means.

File: src/quo2/components/list_items/contact_item.cljs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                               :status-indicator? true
+                               :online?           true
+                               :size              :small
+                               :ring?             false}]
+     [rn/view {:style {:margin-left 8}}
+      [rn/view {:style {:flex-direction :row}}
+       [rn/text {:style (merge typography/paragraph-1 typography/font-semi-bold
+                               {:color (colors/theme-colors colors/neutral-100 colors/white)})}
+        display-name]
+       ; TODO: real name need to be also displayed if set. Currently user model does not have a "real-name" field
```
Comment (ilmotta) [inline src/quo2/components/list_items/contact_item.cljs]: At least in Emacs, this line comment without double semicolons makes the text jump to the end (because of Lisps conventions). Like this:

```clojure
[rn/view {:style {:flex-direction :row}}
       [rn/text {:style (merge typography/paragraph-1 typography/font-semi-bold
                               {:color (colors/theme-colors colors/neutral-100 colors/white)})}
        display-name]
                                        ; TODO: real name need to be also displayed if set. Currently user model does not have a "real-name" field
       (if (:ens-verified item)
         [rn/view {:style {:margin-left 5 :margin-top 4}}
          [icons/icon :main-icons2/verified {:size 12 :color (colors/theme-colors colors/success-50 colors/success-60)}]]
         (when (:added? item)
           [rn/view {:style {:margin-left 5 :margin-top 4}}
            [icons/icon :main-icons2/contact {:size 12 :color (colors/theme-colors colors/primary-50 colors/primary-60)}]]))]
```

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                                :ellipsize-mode      :tail
+                                :accessibility-label :chat-message-text}]
+          :length     0}
+         parsed-text)]
+    (:components result)))
+
+(defn messages-home-item [item]
+  (let [{:keys [chat-id color group-chat last-message timestamp name]} item
+        display-name (if-not group-chat (first (<sub [:contacts/contact-two-names-by-identity chat-id])) name)
+        contact      (if-not group-chat (<sub [:contacts/contact-by-address chat-id]) nil)
```
Comment (ilmotta) [inline src/quo2/components/list_items/messages_home_item.cljs]: I think here `when-not` is sufficient, as it already returns `nil`.

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    [rn/touchable-opacity (merge {:style         {:margin-top         8
+                                                  :margin-horizontal  8
+                                                  :padding-vertical   8
+                                                  :padding-horizontal 12
+                                                  :border-radius      12
+                                                  :flex-direction     :row
+                                                  :align-items        :center}
+                                  :on-press      (fn []
+                                                   (re-frame/dispatch [:dismiss-keyboard])
+                                                   (if platform/android?
```
Comment (ilmotta) [inline src/quo2/components/list_items/messages_home_item.cljs]: Just food for thought (no specific suggestion for change in this PR), but whenever I see a check for `platform/android?` (or any platform check for that matter), I wonder if we should push such checks to the implementation side and not the caller side.

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            [quo2.components.icon :as icons]
+            [quo2.foundations.colors :as colors]
+            [quo2.components.avatars.user-avatar :as user-avatar]
+            [quo.react-native :as rn]
+            [status-im.ui.screens.chat.sheets :as sheets]
+            [quo.platform :as platform]))
+
+(def max-subheader-length 100)
+
+(defn truncate-literal [literal]
```
Comment (ilmotta) [inline src/quo2/components/list_items/messages_home_item.cljs]: I know this function was copied from the old implementation, but I'll just comment it'd be preferable to use the core functions `count` instead of `length` and `subs` instead of `substring`.

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  (when literal
+    (let [size (min max-subheader-length (.-length literal))]
+      {:components (.substring literal 0 size)
+       :length     size})))
+
+(defn add-parsed-to-subheader [acc {:keys [type destination literal children]}]
```
Comment (ilmotta) [inline src/quo2/components/list_items/messages_home_item.cljs]: I understand it's code copied from existing prod code, so I'll just leave a comment without any suggestion for change.

The `add-parsed-to-subheader`, `render-subheader` and `truncate-literal` are begging to be refactored into a pure function outside of the view layer, properly unit tested, and then used in `messages-home-item` namespace to build hiccup. We should be building a dumb view layer as much as possible.

It would be an interesting code challenge :100:

File: src/status_im/ui2/screens/chat/home.cljs [hunk 7]
```diff
@@ -232,9 +195,25 @@
...
 
-(defn contact-requests [count]
+(defn get-display-name [{:keys [chat-id message]}]
+  (let [name        (first (<sub [:contacts/contact-two-names-by-identity chat-id]))
+        no-ens-name (str/blank? (get-in message [:content :ens-name]))]
+    (if no-ens-name
+      (first (str/split name " "))
+      name)))
+
+(defn requests-summary [requests]
```
Comment (ilmotta) [inline src/status_im/ui2/screens/chat/home.cljs:210]: I believe this function should use translations with placeholders, like `hello {{name}}`, or at least not hardcode English words.

File: src/quo2/components/list_items/contact_item.cljs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                               :profile-picture   photo-path
+                               :status-indicator? true
+                               :online?           true
+                               :size              :small
+                               :ring?             false}]
+     [rn/view {:style {:margin-left 8}}
+      [rn/view {:style {:flex-direction :row}}
+       [rn/text {:style (merge typography/paragraph-1 typography/font-semi-bold
+                               {:color (colors/theme-colors colors/neutral-100 colors/white)})}
+        display-name]
```
Comment (yqrashawn) [inline src/quo2/components/list_items/contact_item.cljs]: Just personal preference, I think the code will be easier to review and maintain if put these small meaningful (in design/code level) components into separate small functions. 

Etc `display-name` for this one, it will be easier to add doc/comments to each logical components, since it seems there are lots of UI logic behind this.

**Comment by du82** [comment]
Long messages overflow

File: src/quo2/components/list_items/contact_item.cljs (hunk 1 above)
Comment (OmarBasem) [reply src/quo2/components/list_items/contact_item.cljs]: Actually, I think we agreed on `>evt`.

File: src/quo2/components/list_items/messages_home_item.cljs (hunk 4 above)
Comment (OmarBasem) [reply src/quo2/components/list_items/messages_home_item.cljs]: What do you mean exactly? (I did not write this particluar on-press function, so I am not sure why the `platform` check is needed)

File: src/quo2/components/list_items/messages_home_item.cljs (hunk 6 above)
Comment (OmarBasem) [reply src/quo2/components/list_items/messages_home_item.cljs]: Thanks @ilmotta. I will do that.

File: src/quo2/components/list_items/contact_item.cljs (hunk 8 above)
Comment (OmarBasem) [reply src/quo2/components/list_items/contact_item.cljs]: Thanks @yqrashawn. That's right.

File: src/quo2/components/list_items/messages_home_item.cljs (hunk 4 above)
Comment (ilmotta) [reply src/quo2/components/list_items/messages_home_item.cljs]: I'll try to expand @OmarBasem, but it's definitely not something your PR should attempt to improve necessarily, as this pattern is seen in other parts of the codebase.

Every time we inline platform checks, like `platform/android?` we run the risk of other places that forget to do this check to behave in unexpected ways. For example, for this particular line, if some other part of the code decides to dispatch `:chat.ui/navigate-to-chat-nav2` and forgets to put the platform guard clause, it'll be incorrectly dispatched for iOS. In this way, one could argue that it's the implementation responsibility to decide what to do, and not the caller that should know a hidden implementation detail and decide if it works in iOS or not. Of course there are exceptions.

In other words, when I read `(if platform/android? (dispatch [:A]) (dispatch [:B]))` I'm left wandering, why `:A` only works in Android? It reads to me like a hidden implementation detail leaking to the outside.

IMO what we should be building instead, is an explicit interface for things that behave differently between platforms. There are many other interesting examples in our codebase that would make my point easier to understand. If my examples are not clear enough, let's remember to chat about it next week!

I think we'll get back to this discussion little by little :)

**Review by ilmotta — APPROVED** [review]

File: src/quo2/components/list_items/messages_home_item.cljs (hunk 4 above)
Comment (OmarBasem) [reply src/quo2/components/list_items/messages_home_item.cljs]: Thanks for the extensive explanation @ilmotta.

File: src/quo2/components/list_items/contact_item.cljs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  (if platform/android?
+    (re-frame/dispatch [:chat.ui/navigate-to-chat-nav2 chat-id])
+    (re-frame/dispatch [:chat.ui/navigate-to-chat chat-id]))
+  (re-frame/dispatch [:search/home-filter-changed nil])
+  (re-frame/dispatch [:accept-all-activity-center-notifications-from-chat chat-id]))
+
+(defn contact-item [item]
+  (let [{:keys [public-key]} item
+        display-name (first (<sub [:contacts/contact-two-names-by-identity public-key]))
+        photo-path   (if-not (empty? (:images item)) (<sub [:chats/photo-path public-key]) nil)]
```
Comment (ibrkhalil) [inline src/quo2/components/list_items/contact_item.cljs]: That's the correct sub! 😅

File: src/quo2/components/list_items/contact_item.cljs [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+                               :online?           true
+                               :size              :small
+                               :ring?             false}]
+     [rn/view {:style {:margin-left 8}}
+      [rn/view {:style {:flex-direction :row}}
+       [rn/text {:style (merge typography/paragraph-1 typography/font-semi-bold
+                               {:color (colors/theme-colors colors/neutral-100 colors/white)})}
+        display-name]
+       ; TODO: real name need to be also displayed if set. Currently user model does not have a "real-name" field
+       (if (:ens-verified item)
```
Comment (ibrkhalil) [inline src/quo2/components/list_items/contact_item.cljs]: Can we destructure `:ens-verified` in line 21 ?

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+                                :margin-left 8})}
+        (time/to-short-str (or timestamp "9999999999999"))]] ; placeholder for community chats to avoid crashing until implemented
+      (if (string/blank? (get-in last-message [:content :parsed-text]))
+        [rn/text {:style (merge typography/paragraph-2 typography/font-regular
+                                {:color (colors/theme-colors colors/neutral-50 colors/neutral-40)})}
+         (get-in last-message [:content :text])]
+        [render-subheader (get-in last-message [:content :parsed-text])])]
+     (if (> (:unviewed-mentions-count item) 0)
+       [info-count (:unviewed-mentions-count item) {:top 16}]
+       (when (> (:unviewed-messages-count item) 0)
```
Comment (ibrkhalil) [inline src/quo2/components/list_items/messages_home_item.cljs]: Can we destructure this at line 69

File: src/status_im/ui2/screens/chat/home.cljs [hunk 12]
```diff
@@ -183,7 +146,7 @@
 
 (defn find-contact-requests [notifications]
   (let [received-requests (atom [])
-        has-unread (atom false)]
+        has-unread        (atom false)]
```
Comment (ibrkhalil) [inline src/status_im/ui2/screens/chat/home.cljs]: Can we change it to `has-unread?` ?

**Review by ibrkhalil — APPROVED** [review]

File: src/quo2/components/list_items/contact_item.cljs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            [status-im.utils.handlers :refer [<sub]]
+            [quo2.foundations.typography :as typography]
+            [quo2.components.icon :as icons]
+            [quo2.foundations.colors :as colors]
+            [quo2.components.avatars.user-avatar :as user-avatar]
+            [quo.react-native :as rn]
+            [status-im.utils.utils :refer [get-shortened-address]]
+            [quo.platform :as platform]))
+
+(defn open-chat [chat-id]
```
Comment (flexsurfer) [inline src/quo2/components/list_items/contact_item.cljs]: quo should be abstract from application, no usage of re-frame or app logic

**Review by flexsurfer — CHANGES_REQUESTED** [review]

File: src/quo2/components/list_items/contact_item.cljs [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  (re-frame/dispatch [:dismiss-keyboard])
+  (if platform/android?
+    (re-frame/dispatch [:chat.ui/navigate-to-chat-nav2 chat-id])
+    (re-frame/dispatch [:chat.ui/navigate-to-chat chat-id]))
+  (re-frame/dispatch [:search/home-filter-changed nil])
+  (re-frame/dispatch [:accept-all-activity-center-notifications-from-chat chat-id]))
+
+(defn contact-item [item]
+  (let [{:keys [public-key]} item
+        display-name (first (<sub [:contacts/contact-two-names-by-identity public-key]))
```
Comment (flexsurfer) [inline src/quo2/components/list_items/contact_item.cljs]: same here , no re-frame in quo

File: src/quo2/components/list_items/messages_home_item.cljs [hunk 15]
```diff
@@ -0,0 +1,136 @@
+(ns quo2.components.list-items.messages-home-item
```
Comment (flexsurfer) [inline src/quo2/components/list_items/messages_home_item.cljs]: this shouldn't be in quo namespace, we need to move it to chat components ui

File: src/status_im/ui2/screens/chat/home.cljs [hunk 16]
```diff
@@ -232,9 +198,25 @@
...
-(defn contact-requests [count]
+(defn get-display-name [{:keys [chat-id message]}]
+  (let [name        (first (<sub [:contacts/contact-two-names-by-identity chat-id]))
+        no-ens-name (str/blank? (get-in message [:content :ens-name]))]
+    (if no-ens-name
+      (first (str/split name " "))
+      name)))
+
+(defn requests-summary [requests]
+  (cond
```
Comment (smohamedjavid) [inline src/status_im/ui2/screens/chat/home.cljs]: Can we replace `cond` with `case` (switch) for better performance?

**Review by yqrashawn — APPROVED** [review]

File: src/quo2/components/list_items/received_cr_item.cljs [hunk 17]
```diff
@@ -1,4 +1,4 @@
-(ns quo2.components.list-items.received-contact-request
+(ns quo2.components.list-items.received-cr-item
```
Comment (flexsurfer) [inline src/quo2/components/list_items/received_cr_item.cljs:1]: what's the name of the component in figma?

File: src/quo2/components/notifications/info_count.cljs [hunk 18]
```diff
@@ -0,0 +1,15 @@
+(ns quo2.components.notifications.info-count
+  (:require [quo.react-native :as rn]
+            [quo2.foundations.colors :as colors]
+            [quo2.foundations.typography :as typography]))
+
+(defn info-count [count style]
+  (when (> count 0)
```
Comment (flexsurfer) [inline src/quo2/components/notifications/info_count.cljs:7]: `pos?`

File: src/status_im/ui2/screens/chat/components/contact_item/test.cljs [hunk 19]
```diff
@@ -0,0 +1,3 @@
+(ns status-im.ui2.screens.chat.components.contact-item.test)
+
+;; TODO: tests to be added when RNTL is integrated into our project
```
Comment (flexsurfer) [inline src/status_im/ui2/screens/chat/components/contact_item/test.cljs:3]: so this can be removed for now ?

File: src/status_im/ui2/screens/chat/components/contact_item/view.cljs [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+  (if platform/android?
+    (>evt [:chat.ui/navigate-to-chat-nav2 chat-id])
+    (>evt [:chat.ui/navigate-to-chat chat-id]))
+  (>evt [:search/home-filter-changed nil])
+  (>evt [:accept-all-activity-center-notifications-from-chat chat-id]))
+
+(defn contact-item [item]
+  (let [{:keys [public-key ens-verified added? images]} item
+        display-name (first (<sub [:contacts/contact-two-names-by-identity public-key]))
+        photo-path   (if-not (empty? images) (<sub [:chats/photo-path public-key]) nil)]
```
Comment (flexsurfer) [inline src/status_im/ui2/screens/chat/components/contact_item/view.cljs:24]: `when-not`

File: src/status_im/ui2/screens/chat/components/message_home_item/view.cljs [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+   [text/text {:style (style/timestamp)}
+    (time/to-short-str timestamp)]])
+
+(defn messages-home-item [item]
+  (let [{:keys [chat-id color group-chat last-message timestamp name unviewed-mentions-count unviewed-messages-count]} item
+        display-name (if-not group-chat (first (<sub [:contacts/contact-two-names-by-identity chat-id])) name)
+        contact      (when-not group-chat (<sub [:contacts/contact-by-address chat-id]))
+        photo-path   (when-not (empty? (:images contact)) (<sub [:chats/photo-path chat-id]))]
+    [rn/touchable-opacity (merge {:style         (style/container)
+                                  :on-press      (fn []
```
Comment (flexsurfer) [inline src/status_im/ui2/screens/chat/components/message_home_item/view.cljs:93]: probably better to move it to function outside

**Review by flexsurfer — APPROVED** [review]

**Comment by OmarBasem** [comment]
Hey @VladimrLitvinenko, I see you self-assigned this PR for QA testing. I do not think there is a flow to be QA tested. This PR is new UI implementation.

**Comment by VolodLytvynenko** [comment]
Hi @OmarBasem . Ok, got it
